<a href="https://colab.research.google.com/github/NikoletaEm/datasps/blob/main/07_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**

# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section.

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

In this section, please describe the dataset you are using.  Include a link to the source of this data.  You should also provide some explanation on why you choose this dataset.

https://www.kaggle.com/datasets/bharatnatrayn/movies-dataset-for-feature-extracion-prediction?resource=download

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include:

* Summary statistics means, medians, quartiles,
* Missing value information
* Any other relevant information about the dataset.  



# ================================
## **1. Create DataFrames and Load Data**
# ================================
We'll import pandas for data manipulation and load the dataset directly from the GitHub URL

In [41]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/NikoletaEm/datasps/refs/heads/main/movies.csv"
data = pd.read_csv(url)

# Display dataset information
print("Dataset Information:\n")
print(data.info())

# Display the first 5 rows to understand its structure
print("\nFirst 5 Rows of the Dataset:\n")
print(data.head())

Dataset Information:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9999 non-null   object 
 1   YEAR      9355 non-null   object 
 2   GENRE     9919 non-null   object 
 3   RATING    8179 non-null   float64
 4   ONE-LINE  9999 non-null   object 
 5   STARS     9999 non-null   object 
 6   VOTES     8179 non-null   object 
 7   RunTime   7041 non-null   float64
 8   Gross     460 non-null    object 
dtypes: float64(2), object(7)
memory usage: 703.2+ KB
None

First 5 Rows of the Dataset:

                                MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                      

The dataset contains 9999 rows and 9 columns.
Some columns, such as RATING, VOTES, and RunTime, have significant missing values.
The Gross column is almost entirely missing, which may require removal or special handling.
Columns like YEAR, GENRE, and STARS are text-heavy, suggesting potential for text cleaning or feature engineering.

# ================================
## **2. Summary Statistics**
# ================================
Now we'll examine key statistics such as mean, median, quartiles, and ranges to understand the dataset's distribution.

In [42]:
# Summary statistics for numerical columns
print(data.describe())


            RATING      RunTime
count  8179.000000  7041.000000
mean      6.921176    68.688539
std       1.220232    47.258056
min       1.100000     1.000000
25%       6.200000    36.000000
50%       7.100000    60.000000
75%       7.800000    95.000000
max       9.900000   853.000000


- The "RATING" column shows a mean value of 6.92, with a standard deviation of 1.22.This indicates that the ratings are generally clustered around the mean, with a typical variation of about 1.22 points.
- The "RunTime" column has a much larger standard deviation (47.26), meaning that the movie runtimes vary significantly from the mean value of 68.69 minutes. The presence of a maximum value of 853 minutes suggests that there might be an outlier or erroneous data point in this column.
- The 25th, 50th (median), and 75th percentiles give insight into the distribution of the data and show that the majority of movies have runtimes between 36 and 95 minutes.

# ================================
## **3. Missing Values Information**
# ================================

In [43]:
# Missing values
print(data.isnull().sum())

MOVIES         0
YEAR         644
GENRE         80
RATING      1820
ONE-LINE       0
STARS          0
VOTES       1820
RunTime     2958
Gross       9539
dtype: int64


- The "YEAR" column has 644 missing entries, which could potentially be critical since it represents the release year of the movie. We may need to explore imputation strategies based on other columns(e.g., genre, box office gross) or consider removing rows with missing years if necessary.
- Both the "RATING" and "VOTES" columns have 1820 missing values, which indicates a significant amount of missing data for the movie ratings and votes. Since these are important for understanding the movie's reception, we may need to address this by either filling in the missing values or removing rows.
- The "RunTime" column has 2958 missing values, which is also a substantial portion of the data.Depending on the analysis, we might consider removing these rows or filling them with a reasonable default value (e.g., the median runtime).
- The "Gross" column has the most missing data, with 9539 missing entries, which is nearly the entire column.This may require a more detailed review of whether it's possible to impute these values or if this column an be excluded from certain analyses.


# ================================
## **4.Identifying Outliers**
# ================================

Detecting outliers can be important to ensure the quality of your data. Outliers can skew statistical analyses. We can use methods like the Interquartile Range (IQR) to detect and potentially remove outliers.



In [44]:
Q1 = data['RunTime'].quantile(0.25)
Q3 = data['RunTime'].quantile(0.75)
IQR = Q3 - Q1

# Calculate lower and upper bounds for outliers
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR


# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort your data based on multiple variables.

8. Filter your data based on some condition.

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group.
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups.

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**


# ================================
##**Create a Subset of the Original Data**
# ================================

We will create a subset of the original data by selecting specific columns.This helps in focusing on certain columns of interest for analysis or modification.


In [87]:
subset_data = data[['MOVIES', 'YEAR', 'RATING', 'RunTime', 'GENRE','ONE-LINE']]

print("Subset of the data (first 5 rows):")
print(subset_data.head())

Subset of the data (first 5 rows):
                                MOVIES         YEAR  RATING  RunTime  \
0                        Blood Red Sky       (2021)     6.1    121.0   
1  Masters of the Universe: Revelation     (2021– )     5.0     25.0   
2                     The Walking Dead  (2010–2022)     8.2     44.0   
3                       Rick and Morty     (2013– )     9.2     23.0   
4                      Army of Thieves       (2021)     NaN      NaN   

                                        GENRE  \
0      \nAction, Horror, Thriller               
1  \nAnimation, Action, Adventure               
2       \nDrama, Horror, Thriller               
3  \nAnimation, Adventure, Comedy               
4         \nAction, Crime, Horror               

                                            ONE-LINE  
0  \nA woman with a mysterious illness is forced ...  
1  \nThe war for Eternia begins again in what may...  
2  \nSheriff Deputy Rick Grimes wakes up from a c...  
3  \nAn animated 

# ================================
## **1. Modify Multiple Column Names**
# ================================

Let's modify the column names to be more consistent and easier to work with.
For example, we can convert them to lower case. Additionally, we'll rename 'RunTime' to 'run_time'.


In [88]:
subset_data.columns = subset_data.columns.str.lower()
subset_data.rename(columns={'runtime': 'run_time'}, inplace=True)

print("\nColumn names after modification:")
print(subset_data.columns)


Column names after modification:
Index(['movies', 'year', 'rating', 'run_time', 'genre', 'one-line'], dtype='object')


<ipython-input-88-2e3a99bbcfb3>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_data.rename(columns={'runtime': 'run_time'}, inplace=True)


# ================================
## **2. Look at the Structure of the Data**
# ================================

We need to check if any variables are improperly coded (e.g., strings or characters that should be numerical)


In [89]:
print("\nData types before conversion:")
print(subset_data.dtypes)


Data types before conversion:
movies       object
year         object
rating      float64
run_time    float64
genre        object
one-line     object
dtype: object


- movies: It's an object, which is fine as it's a string column.
- year: It's an object, which is fine since it's representing years in the form of text (e.g., "2021" or "2010–2022").
- rating: This is already a float (float64), so no changes are needed here.
- run_time: This is a float (float64), which is correct as it's numerical data representing runtime in minutes.
- genre: It's an object (string), which is fine as it represents categorical data for genres.

In [90]:
# Fill missing values for 'rating' with the median using .loc
subset_data.loc[:, 'rating'] = subset_data['rating'].fillna(median_rating)

# Extract and convert 'year' to integers, then fill missing values with the median year
subset_data.loc[:, 'year'] = pd.to_numeric(subset_data['year'].str.extract(r'(\d{4})')[0], errors='coerce')
subset_data.loc[:, 'year'] = subset_data['year'].fillna(median_year)

# Fill missing values in 'run_time' with the median runtime
subset_data.loc[:, 'run_time'] = subset_data['run_time'].fillna(median_runtime)

# Fill missing values in 'genre' with 'Unknown'
subset_data.loc[:, 'genre'] = subset_data['genre'].fillna('Unknown')


<ipython-input-90-c69446eefb1c>:6: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  subset_data.loc[:, 'year'] = subset_data['year'].fillna(median_year)


In [92]:
print(subset_data.isnull().sum())

movies      0
year        0
rating      0
run_time    0
genre       0
one-line    0
dtype: int64


 **Create New Columns Based on Existing Columns**

In [94]:
# Extract decade by converting year to int and grouping by tens
subset_data.loc[:, 'decade'] = (subset_data['year'].astype(int) // 10) * 10

subset_data['rating_category'] = pd.cut(subset_data['rating'], bins=[0, 5, 7, 10], labels=['Low', 'Medium', 'High'])

# Check the first few rows to confirm the new column
print(subset_data.head())

                                movies    year  rating  run_time  \
0                        Blood Red Sky  2021.0     6.1     121.0   
1  Masters of the Universe: Revelation  2021.0     5.0      25.0   
2                     The Walking Dead  2010.0     8.2      44.0   
3                       Rick and Morty  2013.0     9.2      23.0   
4                      Army of Thieves  2021.0     7.1      60.0   

                                        genre  \
0      \nAction, Horror, Thriller               
1  \nAnimation, Action, Adventure               
2       \nDrama, Horror, Thriller               
3  \nAnimation, Adventure, Comedy               
4         \nAction, Crime, Horror               

                                            one-line  decade RATING_CATEGORY  \
0  \nA woman with a mysterious illness is forced ...    2020          Medium   
1  \nThe war for Eternia begins again in what may...    2020             Low   
2  \nSheriff Deputy Rick Grimes wakes up from a c...    

**Let's drop the ONE-LINE column since it seems less useful for our analysis.**

In [95]:
# Drop 'ONE-LINE' column
subset_data = subset_data.loc[:, subset_data.columns != 'one-line']



**Drop Rows Where 'genre' is 'Unknown'**

---



In [96]:
subset_data = subset_data[subset_data['genre'] != 'Unknown']

**We'll sort movies first by rating (highest to lowest) and then by year (most recent first).**

In [97]:
subset_data = subset_data.sort_values(by=['rating', 'year'], ascending=[False, False])


**We'll filter for movies with a rating above 8 and a runtime over 100 minutes.**

In [98]:
# Filter for movies with high ratings and long runtimes
high_rated_long_movies = subset_data[(subset_data['rating'] > 8) & (subset_data['run_time'] > 100)]


**We'll convert all values in the MOVIES column to uppercase.**


In [99]:
# Convert 'MOVIES' column to uppercase
subset_data['movies'] = subset_data['movies'].str.upper()

**Check for Numeric Values in a Column**

In [100]:
non_numeric_decade = subset_data[~subset_data['decade'].apply(pd.to_numeric, errors='coerce').notna()]
print("\nRows where 'decade' has non-numeric values:")
print(non_numeric_decade)


Rows where 'decade' has non-numeric values:
Empty DataFrame
Columns: [movies, year, rating, run_time, genre, decade, RATING_CATEGORY, rating_category]
Index: []


All its rows contain numeric values that's good!

**Group Your Dataset by One Column and Get the Mean, Min, and Max Values by Group**


In [101]:
# Group by 'GENRE' and calculate mean, min, and max values for 'RATING'
grouped_data = subset_data.groupby('genre')['rating'].agg(['mean', 'min', 'max'])

# Display the grouped data
print(grouped_data)

                                                mean  min  max
genre                                                         
\nAction                                    5.769444  3.1  7.8
\nAction, Adventure                         5.980000  4.0  7.1
\nAction, Adventure, Biography              7.725000  6.9  8.5
\nAction, Adventure, Comedy                 6.913761  2.7  9.2
\nAction, Adventure, Crime                  6.376786  2.1  7.9
...                                              ...  ...  ...
\nTalk-Show, Sport                          7.100000  7.1  7.1
\nThriller                                  5.876923  3.1  8.2
\nThriller, Mystery                         7.100000  7.1  7.1
\nWar                                       8.200000  8.2  8.2
\nWestern                                   6.650000  6.2  7.1

[510 rows x 3 columns]


**Group by Two Columns and Then Sort the Aggregated Results Within the Groups**

# Conclusions  

After exploring your dataset, provide a short summary of what you noticed from this dataset.  What would you explore further with more time?

In [102]:
# Group by both 'GENRE' and 'YEAR' and get mean 'RATING', then sort within each group
grouped_sorted_data = subset_data.groupby(['genre', 'year'])['rating'].agg(['mean']).sort_values(by=['genre', 'year'], ascending=[True, False])

# Display the sorted and grouped data
print(grouped_sorted_data)


                                  mean
genre                 year            
\nAction              2021.0  7.100000
                      2020.0  5.250000
                      2019.0  4.733333
                      2018.0  6.818182
                      2017.0  5.525000
...                                ...
\nWestern             2018.0  7.100000
                      1955.0  6.900000
                      1954.0  6.700000
                      1953.0  6.400000
                      1952.0  6.400000

[2194 rows x 1 columns]


**Outlier Identification and Handling**

In [104]:
outlier_counts = {}

# Loop through each numeric column in subset_data
for column in subset_data.select_dtypes(include=['number']).columns:
    # Calculate IQR for the column
    Q1 = subset_data[column].quantile(0.25)
    Q3 = subset_data[column].quantile(0.75)
    IQR = Q3 - Q1

    # Calculate lower and upper bounds for outliers
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Identify outliers for the column
    outliers = subset_data[(subset_data[column] < lower_bound) | (subset_data[column] > upper_bound)]

    # Count the number of outliers
    outlier_count = outliers.shape[0]

    # Store the outlier count in the dictionary
    outlier_counts[column] = outlier_count

    # Handle outliers: Replacing outliers with the median of the column
    subset_data[column] = subset_data[column].apply(
        lambda x: subset_data[column].median() if x < lower_bound or x > upper_bound else x
    )

# Print out the number of outliers in each column
for column, count in outlier_counts.items():
    print(f"Column '{column}' has {count} outliers.")

Column 'rating' has 452 outliers.
Column 'run_time' has 44 outliers.
Column 'decade' has 0 outliers.
